src/main/java/software/amazon/smithy/lsp/SmithyLanguageServer.java

```diff
@@ -443,40 +468,54 @@ public void didChangeWatchedFiles(DidChangeWatchedFilesParams params) {
-        try {
-            client.createProgress(new WorkDoneProgressCreateParams(progressToken)).get();
-        } catch (ExecutionException | InterruptedException e) {
-            client.error(String.format("Unable to create work done progress token: %s", e.getMessage()));
-            progressToken = null;
+        for (WorkspaceFolder folder : params.getEvent().getAdded()) {
+            loadWorkspace(folder);
         }
 
-        if (progressToken != null) {
-            WorkDoneProgressBegin begin = new WorkDoneProgressBegin();
-            begin.setTitle("Updating workspace");
-            client.notifyProgress(new ProgressParams(progressToken, Either.forLeft(begin)));
+        for (WorkspaceFolder folder : params.getEvent().getRemoved()) {
+            removeWorkspace(folder);
         }
 
-        for (WorkspaceFolder folder : params.getEvent().getAdded()) {
-            Path root = Paths.get(URI.create(folder.getUri()));
-            tryInitProject(folder.getName(), root);
-        }
+        unregisterSmithyFileWatchers().thenRun(this::registerSmithyFileWatchers);
+        unregisterWorkspaceBuildFileWatchers().thenRun(this::registerWorkspaceBuildFileWatchers);
+        sendFileDiagnosticsForManagedDocuments();
+    }
 
-        for (WorkspaceFolder folder : params.getEvent().getRemoved()) {
-            Project removedProject = projects.removeProjectByName(folder.getName());
-            if (removedProject == null) {
-                continue;
+    private void loadWorkspace(WorkspaceFolder workspaceFolder) {
+        Path workspaceRoot = Paths.get(URI.create(workspaceFolder.getUri()));
+        workspacePaths.add(workspaceRoot);
+        try {
+            List<Path> projectRoots = ProjectRootVisitor.findProjectRoots(workspaceRoot);
+            for (Path root : projectRoots) {
+                tryInitProject(root);
             }
+        } catch (IOException e) {
+            LOGGER.severe(e.getMessage());
+        }
+    }
 
-            resolveDetachedProjects(removedProject, Project.empty(removedProject.root()));
+    private void removeWorkspace(WorkspaceFolder folder) {
+        Path workspaceRoot = Paths.get(URI.create(folder.getUri()));
+        workspacePaths.remove(workspaceRoot);
+
+        // Have to do the removal separately, so we don't modify project.attachedProjects()
+        // while iterating through it
+        List<String> projectsToRemove = new ArrayList<>();
+        for (var entry : projects.attachedProjects().entrySet()) {
+            if (entry.getValue().root().startsWith(workspaceRoot)) {
+                projectsToRemove.add(entry.getKey());
+            }
         }
 
-        unregisterSmithyFileWatchers().thenRun(this::registerSmithyFileWatchers);
-        sendFileDiagnosticsForManagedDocuments();
+        for (String projectName : projectsToRemove) {
+            removeProjectAndResolveDetached(projectName);
```

Comment:
do we need to do this in a separate loop from the above? Could we just remove as we go?

Reply:
There's a comment above the above loop:
```
        // Have to do the removal separately, so we don't modify project.attachedProjects()
        // while iterating through it
```
Removing modifies the attachedProjects map, which is what backs the entrySet being iterated over